alacritty_terminal/src/tty/windows/conpty.rs

```diff
@@ -79,63 +82,92 @@ impl ConptyApi {
     }
 }
 
-/// RAII Pseudoconsole
+/// RAII Pseudoconsole. Closes pipes and Pty when dropped.
 pub struct Conpty {
-    pub handle: HPCON,
-    api: ConptyApi,
+    inner: Arc<ConptyInner>,
+    pub conout: EventedReadablePipe,
+    pub conin: EventedWritablePipe,
 }
 
-/// Handle can be cloned freely and moved between threads.
-pub type ConptyHandle = Arc<Conpty>;
+impl Conpty {
+    pub fn new(size: &SizeInfo) -> Option<Self> {
+        let api = ConptyApi::new()?;
+
+        let mut handle = 0 as HPCON;
+
+        // Passing 0 as the size parameter allows the "system default" buffer
+        // size to be used. There may be small performance and memory advantages
+        // to be gained by tuning this in the future, but it's likely a reasonable
+        // start point.
+        let (conout, conout_pty_handle) = miow::pipe::anonymous(0).unwrap();
+        let (conin_pty_handle, conin) = miow::pipe::anonymous(0).unwrap();
+
+        let coord =
+            coord_from_sizeinfo(size).expect("Overflow when creating initial size on pseudoconsole");
+
+        // Create the Pseudo Console, using the pipes
+        let result = unsafe {
+            (api.CreatePseudoConsole)(
+                coord,
+                conin_pty_handle.into_raw_handle(),
+                conout_pty_handle.into_raw_handle(),
+                0,
+                &mut handle as *mut HPCON,
+            )
+        };
+
+        assert_eq!(result, S_OK);
+
+        let inner = ConptyInner { handle, api, closed: AtomicBool::new(false) };
+
+        Some(Self {
+            inner: Arc::new(inner),
+            conout: EventedReadablePipe::Anonymous(EventedAnonRead::new(conout)),
+            conin: EventedWritablePipe::Anonymous(EventedAnonWrite::new(conin)),
+        })
+    }
+
+    pub fn resize_handle(&self) -> ConptyResizeHandle {
+        ConptyResizeHandle(self.inner.clone())
+    }
+}
+
+/// Resize handle can be moved between threads and outlive the ConPTY safely
+pub struct ConptyResizeHandle(Arc<ConptyInner>);
+
+pub struct ConptyInner {
+    handle: HPCON,
+    api: ConptyApi,
+    closed: AtomicBool,
+}
 
 impl Drop for Conpty {
     fn drop(&mut self) {
-        unsafe { (self.api.ClosePseudoConsole)(self.handle) }
```

Comment:
Would it not be possible to implement `Drop` for `PTY` and upholding the drop order in there? I still absolutely cannot stand that we move the `conin` and `conout` on the WinPTY/ConPTY for no good reason at all. All we're creating here is duplication of code that will come back to haunt us in the future.

Reply:
> Would it not be possible to implement Drop for PTY and upholding the drop order in there? I still absolutely cannot stand that we move the conin and conout on the WinPTY/ConPTY for no good reason at all.

I think I'm failing to explain the deadlock very well. The [docs for `ClosePseudoConsole`](https://docs.microsoft.com/en-us/windows/console/closepseudoconsole) describe what I believe is the issue:

> A final painted frame may arrive on hOutput from the pseudoconsole when this API is called. It is expected that the caller will drain this information from the communication channel buffer and either present it or discard it. Failure to drain the buffer may cause the Close call to wait indefinitely until it is drained or the communication channels are broken another way.

So we need to ensure that the `output` pipe is still open when `ClosePseudoConsole` is called, and also we need to then continuously read from it until it's empty. This will have to be done in another thread, because the main thread will be blocked by `ClosePseudoConsole`. The `EventedAnonRead` type actually uses a reader thread and an internal buffer, so this is solved for us as long as `conout` is still alive when `ClosePseudoConsole` is called.

Going back to this point:

> Would it not be possible to implement Drop for PTY and upholding the drop order in there?

Yes, we could definitely do that. It would lead to less code. However:

- From my observations, we ***need*** to call `ClosePseudoConsole` or the pseudoconsole is leaked after alacritty closes. If it leaks, the conhost pseudoconsole process (and child shell/process tree) becomes a system process which never closes until the computer shuts down.
- My worry is that if `ClosePseudoConsole` call isn't in `Conpty`'s drop implementation, then we are dependent on other code explicitly closing the pseudoconsole to avoid these leaks. (i.e. I'm doing my best to stick to RAII for the `Conpty` opening & closing pseudoconsoles.)

The frustration is this potential deadlock in `ClosePseudoConsole`. To ensure that `conout` is still alive when `ClosePseudoConsole` is called, and to stick to RAII and call `ClosePseudoConsole` as part of `Drop for Conpty`, it was necessary to give `Conpty` ownership of the `conout` pipe.

I don't love that it leads to more code either, but I think that making the `Conpty` type robust is worth the price.

I'm ok if you prefer shorter code and instead accept that misuse of the `Conpty` type might cause deadlocks in future, I'll rework this PR to just `impl Drop for Pty` and be done with it 😄

Reply:
> My worry is that if ClosePseudoConsole call isn't in Conpty's drop implementation, then we are dependent on other code explicitly closing the pseudoconsole to avoid these leaks. (i.e. I'm doing my best to stick to RAII for the Conpty opening & closing pseudoconsoles.)

If we implement the drop on ConPTY and call `ClosePseudoConsole` there, we should be safe. The only thing we need to make sure of then is that we do not free the other code before it. But we do not risk leaking anything.

After that is ensured, the only thing we need to check is that the `ConPTY` handle is dropped last in the `Pty` drop implementation. We could ensure that simply by explicitly dropping the handle first and letting Rust automatically close conout/conin (we could also explicitly close those afterwards).

As far as I can tell, this would make everyone drop the things it needs to have 'knowledge' about, while at the same time we still only have a single point where we would need to ensure ordering. I don't see right now how there's difference between ensuring this in the drop for Pty vs ConPTY, please let me know if there is any.

The additional benefit of this strategy would be, that if there is a ConPTY2 some time in the future, we're already on the safe side, because we make sure our `conout`/`conin` are always dropped last, so we will drain the buffers.

> I'm ok if you prefer shorter code and instead accept that misuse of the Conpty type might cause leaks/deadlocks in future.

I don't see how the misuse of the `ConPTY` type could cause leaks here, assuming we close the PTY in its `Drop` implementation. It would be possible to cause a deadlock if we rewrite the `PTY` module and forget about it, but the same deadlock would appear if we rewrite the `ConPTY` module in your PR and forget about it. So as far as I can tell the issue is still the same, just on a different level.

I guess I just don't see what we gain by rewriting things, other than making the code more complicated? There's no doubt the Windows code is already extremely messy and difficult to maintain, so I'd like to stay away from making it even harder.